### ISSUE TITLE
Invalid instruction

### ISSUE THREAD

**Issue by otavio** [lipanski/mockito#60]
Hello,

Using 0.14 and 0.15 I get following backtrace:

```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error("EOF while parsing a value", line: 1, column: 0)', src/libcore/result.rs:997:5
stack backtrace:
test states::download::test::download_has_transition_callback_trait ... ok
   0:     0x55cb79e4a483 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::he292500cb069592b
                               at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1:     0x55cb79e451b8 - std::sys_common::backtrace::_print::h0a244112d948cad7
                               at src/libstd/sys_common/backtrace.rs:70
   2:     0x55cb79e48252 - std::panicking::default_hook::{{closure}}::hb199339b212075e6
                               at src/libstd/sys_common/backtrace.rs:58
                               at src/libstd/panicking.rs:200
   3:     0x55cb79e47fc4 - std::panicking::default_hook::h0f5d79d3aeb5cebe
                               at src/libstd/panicking.rs:215
   4:     0x55cb79e48960 - std::panicking::rust_panic_with_hook::hebcac80ec729c98d
                               at src/libstd/panicking.rs:478
   5:     0x55cb79e484e1 - std::panicking::continue_panic_fmt::h83729debd9444082
                               at src/libstd/panicking.rs:385
   6:     0x55cb79e483c5 - rust_begin_unwind
                               at src/libstd/panicking.rs:312
   7:     0x55cb79e65b2c - core::panicking::panic_fmt::hc3bc1d4b55face00
                               at src/libcore/panicking.rs:85
   8:     0x55cb79637365 - core::result::unwrap_failed::h5f7b56502d8e7619
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/macros.rs:16
   9:     0x55cb796354d5 - <core::result::Result<T, E>>::unwrap::h30bf3792771bba7b
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/result.rs:798
  10:     0x55cb7962e058 - mockito::Matcher::matches_value::h6b88e003f3f163fd
                               at /home/otavio/.cargo/registry/src/github.com-1ecc6299db9ec823/mockito-0.15.0/src/lib.rs:542
  11:     0x55cb7962daf5 - mockito::server::<impl mockito::Mock>::body_matches::hfa524c6f2077ff58
                               at /home/otavio/.cargo/registry/src/github.com-1ecc6299db9ec823/mockito-0.15.0/src/server.rs:25
  12:     0x55cb7962db71 - mockito::server::<impl core::cmp::PartialEq<mockito::request::Request> for &'a mut mockito::Mock>::eq::hfe00a18b02575a17
                               at /home/otavio/.cargo/registry/src/github.com-1ecc6299db9ec823/mockito-0.15.0/src/server.rs:34
  13:     0x55cb7962d173 - core::cmp::impls::<impl core::cmp::PartialEq<&'b B> for &'a A>::eq::h860d9b6762530b92
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/cmp.rs:1003
  14:     0x55cb7963b24a - mockito::server::handle_match_mock::{{closure}}::h43cd654f00ccbb66
                               at /home/otavio/.cargo/registry/src/github.com-1ecc6299db9ec823/mockito-0.15.0/src/server.rs:134
  15:     0x55cb79660850 - core::iter::traits::DoubleEndedIterator::rfind::{{closure}}::h2d6c58c85061bda0
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/iter/traits.rs:635
  16:     0x55cb79625204 - <core::slice::IterMut<'a, T> as core::iter::traits::DoubleEndedIterator>::try_rfold::ha97d61d3f6ae9f27
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/slice/mod.rs:2988
  17:     0x55cb79625897 - core::iter::traits::DoubleEndedIterator::rfind::hd806c94572edc275
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/iter/traits.rs:634
  18:     0x55cb796615dd - <core::iter::Rev<I> as core::iter::iterator::Iterator>::find::h1c9b2002948692d9
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/iter/mod.rs:441
  19:     0x55cb7963adea - mockito::server::handle_match_mock::h9ab341bc060c8b8b
                               at /home/otavio/.cargo/registry/src/github.com-1ecc6299db9ec823/mockito-0.15.0/src/server.rs:134
  20:     0x55cb7963ac0a - mockito::server::handle_request::h1be4e3f2b46e1cf4
                               at /home/otavio/.cargo/registry/src/github.com-1ecc6299db9ec823/mockito-0.15.0/src/server.rs:126
  21:     0x55cb7963a513 - mockito::server::try_start::{{closure}}::h19e1783d6319ac32
                               at /home/otavio/.cargo/registry/src/github.com-1ecc6299db9ec823/mockito-0.15.0/src/server.rs:110
  22:     0x55cb7964cada - std::sys_common::backtrace::__rust_begin_short_backtrace::h360cdcc582893ddf
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/sys_common/backtrace.rs:135
  23:     0x55cb7965c67d - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::hc57d6fa74e4f46d3
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/mod.rs:469
  24:     0x55cb7965b3fd - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h7de1a177b1a95085
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panic.rs:309
  25:     0x55cb7965c844 - std::panicking::try::do_call::h1c0328330a706679
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panicking.rs:297
  26:     0x55cb79e52449 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:92
  27:     0x55cb7965c765 - std::panicking::try::h1d902e6b2af3dd1f
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panicking.rs:276
  28:     0x55cb7965bbcd - std::panic::catch_unwind::haf4142276f498d7e
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panic.rs:388
  29:     0x55cb7965c47d - std::thread::Builder::spawn_unchecked::{{closure}}::ha27e096332f07b6c
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/mod.rs:468
  30:     0x55cb79660e99 - <F as alloc::boxed::FnBox<A>>::call_box::h1270d75b0ae800e1
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/liballoc/boxed.rs:734
  31:     0x55cb79e5180d - std::sys::unix::thread::Thread::new::thread_start::hd9992a1ec79303ca
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/liballoc/boxed.rs:744
                               at src/libstd/sys_common/thread.rs:14
                               at src/libstd/sys/unix/thread.rs:81
  32:     0x7f046884ea9c - start_thread
  33:     0x7f0468762b22 - clone
  34:                0x0 - <unknown>
```

**Comment by otavio** [comment]
I did a `git bisect` and it seems the culprit is:

```
commit 5038d61bad52ab1cafe552aa8bc40f6cf9d2f104
Author: sterlingjensen <5555776+sterlingjensen@users.noreply.github.com>
Date:   Sun Nov 18 06:38:56 2018 -0600

    Fix Windows build and doctest failures. (#51)

    Fixes #41 and #36

    Substitute [http_muncher] with [httparse]
```

It was introduced on #51.

**Comment by otavio** [comment]
The error happens here:

https://github.com/lipanski/mockito/blob/459049a780c49e8837c77cba047b76d5606b322e/src/lib.rs#L541-L544

**Comment by otavio** [comment]
And now full debug output:

```
test client::tests::probe_requirements ... DEBUG - Server is listening at 127.0.0.1:1234
DEBUG - Mock::create() called for
POST /upgrades
content-type: application/json
api-content-type: application/vnd.updatehub-v1+json
{"device-attributes":{"attr1":["attrvalue1"],"attr2":["attrvalue2"]},"device-identity":{"id1":["value1"],"id2":["value2"]},"hardware":"board","product-uid":"229ffd7e08721d716163fc81a2dbaf6c90d449f0a3b009b6a2defe8a0b0d7381","version":"1.1"}

TRACE - registering with poller
TRACE - event Readable Token(4194303)
TRACE - loop process - 1 events, 0.000s
TRACE - checkout waiting for idle connection: "http://127.0.0.1:1234"
TRACE - Http::connect; scheme=http, host=127.0.0.1, port=Some(1234)
DEBUG - connecting to 127.0.0.1:1234
TRACE - registering with poller
TRACE - event Writable Token(0)
TRACE - loop process - 1 events, 0.000s
TRACE - client handshake HTTP/1
TRACE - event Readable Token(4194303)
TRACE - loop process - 1 events, 0.000s
TRACE - read_keep_alive; is_mid_message=false
TRACE - signal: Want
TRACE - signal found waiting giver, notifying
TRACE - flushed({role=client}): State { reading: Init, writing: Init, keep_alive: Busy, error: None }
TRACE - event Readable Token(4194303)
TRACE - loop process - 1 events, 0.000s
TRACE - poll_want: taker wants!
TRACE - event Readable Token(4194303)
TRACE - loop process - 1 events, 0.000s
TRACE - read_keep_alive; is_mid_message=false
TRACE - Client::encode method=POST, body=Some(Known(239))
TRACE - detected no usage of vectored write, flattening
DEBUG - flushed 235 bytes
DEBUG - Request received:
DEBUG - flushed 239 bytes
POST /upgrades
user-agent: updatehub/next
accept: */*
TRACE - flushed({role=client}): State { reading: Init, writing: KeepAlive, keep_alive: Busy, error: None }
content-type: application/json
api-content-type: application/vnd.updatehub-v1+json
api-retries: 0
content-length: 239
accept-encoding: gzip
host: 127.0.0.1:1234
TRACE - loop process - 0 events, 0.000s

body []
Object({"device-attributes": Object({"attr1": Array([String("attrvalue1")]), "attr2": Array([String("attrvalue2")])}), "device-identity": Object({"id1": Array([String("value1")]), "id2": Array([String("value2")])}), "hardware": String("board"), "product-uid": String("229ffd7e08721d716163fc81a2dbaf6c90d449f0a3b009b6a2defe8a0b0d7381"), "version": String("1.1")}) vs ""
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error("EOF while parsing a value", line: 1, column: 0)', src/libcore/result.rs:997:5
stack backtrace:
   0:     0x55a08c275cb3 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::he292500cb069592b
                               at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1:     0x55a08c2709e8 - std::sys_common::backtrace::_print::h0a244112d948cad7
                               at src/libstd/sys_common/backtrace.rs:70
   2:     0x55a08c273a82 - std::panicking::default_hook::{{closure}}::hb199339b212075e6
                               at src/libstd/sys_common/backtrace.rs:58
                               at src/libstd/panicking.rs:200
   3:     0x55a08c2737f4 - std::panicking::default_hook::h0f5d79d3aeb5cebe
                               at src/libstd/panicking.rs:215
   4:     0x55a08c274190 - std::panicking::rust_panic_with_hook::hebcac80ec729c98d
                               at src/libstd/panicking.rs:478
   5:     0x55a08c273d11 - std::panicking::continue_panic_fmt::h83729debd9444082
                               at src/libstd/panicking.rs:385
   6:     0x55a08c273bf5 - rust_begin_unwind
                               at src/libstd/panicking.rs:312
   7:     0x55a08c29138c - core::panicking::panic_fmt::hc3bc1d4b55face00
                               at src/libcore/panicking.rs:85
   8:     0x55a08ba68015 - core::result::unwrap_failed::h2f6a389e7f76f094
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/macros.rs:16
   9:     0x55a08ba65ce5 - <core::result::Result<T, E>>::unwrap::h14c6b2013f146f60
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/result.rs:798
  10:     0x55a08ba7e51c - mockito::Matcher::matches_value::heea7f4edd2ba76ac
                               at /home/otavio/src/mockito/src/lib.rs:543
  11:     0x55a08ba7decc - mockito::server::<impl mockito::Mock>::body_matches::h6a17812c7d421b5f
                               at /home/otavio/src/mockito/src/server.rs:26
  12:     0x55a08ba7df61 - mockito::server::<impl core::cmp::PartialEq<mockito::request::Request> for &'a mut mockito::Mock>::eq::h0c4dc02b8e3f08bb
                               at /home/otavio/src/mockito/src/server.rs:36
  13:     0x55a08ba5f783 - core::cmp::impls::<impl core::cmp::PartialEq<&'b B> for &'a A>::eq::h818070ce41ed6e75
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/cmp.rs:1003
  14:     0x55a08ba7250a - mockito::server::handle_match_mock::{{closure}}::h001e306d3bfb8ce6
                               at /home/otavio/src/mockito/src/server.rs:139
  15:     0x55a08ba734c0 - core::iter::traits::DoubleEndedIterator::rfind::{{closure}}::h2d96be39e1723694
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/iter/traits.rs:635
  16:     0x55a08ba563d4 - <core::slice::IterMut<'a, T> as core::iter::traits::DoubleEndedIterator>::try_rfold::h0d2a97d2182cca8c
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/slice/mod.rs:2988
  17:     0x55a08ba56a67 - core::iter::traits::DoubleEndedIterator::rfind::ha80cd623b445eebe
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/iter/traits.rs:634
  18:     0x55a08ba53c4d - <core::iter::Rev<I> as core::iter::iterator::Iterator>::find::hac541bbae670b146
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libcore/iter/mod.rs:441
  19:     0x55a08ba4fd9a - mockito::server::handle_match_mock::h09852b4ba4774b9b
                               at /home/otavio/src/mockito/src/server.rs:139
  20:     0x55a08ba4fbba - mockito::server::handle_request::hf9c5ac7641b5bf1f
                               at /home/otavio/src/mockito/src/server.rs:131
  21:     0x55a08ba71e23 - mockito::server::try_start::{{closure}}::h09d78db9b477893e
                               at /home/otavio/src/mockito/src/server.rs:113
  22:     0x55a08ba7933a - std::sys_common::backtrace::__rust_begin_short_backtrace::hd8989c585b8337f2
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/sys_common/backtrace.rs:135
  23:     0x55a08ba5116d - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::h02a4feba886f2e1b
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/mod.rs:469
  24:     0x55a08ba5527d - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hab5f69938653e3e1
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panic.rs:309
  25:     0x55a08ba68984 - std::panicking::try::do_call::h918be826bcb5f11a
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panicking.rs:297
  26:     0x55a08c27dca9 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:92
  27:     0x55a08ba688a5 - std::panicking::try::heb2e6cedf3a06f4f
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panicking.rs:276
  28:     0x55a08ba552ad - std::panic::catch_unwind::hf44bf6f2888a839a
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/panic.rs:388
  29:     0x55a08ba50f5d - std::thread::Builder::spawn_unchecked::{{closure}}::hb89ad63ae40423cd
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/mod.rs:468
  30:     0x55a08ba51219 - <F as alloc::boxed::FnBox<A>>::call_box::hdc94455ed67e027c
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/liballoc/boxed.rs:734
  31:     0x55a08c27d06d - std::sys::unix::thread::Thread::new::thread_start::hd9992a1ec79303ca
                               at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/liballoc/boxed.rs:744
                               at src/libstd/sys_common/thread.rs:14
                               at src/libstd/sys/unix/thread.rs:81
  32:     0x7fce7acb3a9c - start_thread
  33:     0x7fce7abc7b22 - clone
  34:                0x0 - <unknown>
TRACE - event Readable | Writable | Error | Hup Token(0)
TRACE - loop process - 1 events, 0.000s
TRACE - Conn::read_head
TRACE - State::close_read()
DEBUG - parse error (an IO error occurred: Connection reset by peer (os error 104)) with 0 bytes
DEBUG - read_head error: an IO error occurred: Connection reset by peer (os error 104)
TRACE - read_keep_alive; is_mid_message=true
TRACE - State::close()
TRACE - flushed({role=client}): State { reading: Closed, writing: Closed, keep_alive: Disabled, error: None }
TRACE - shut down IO complete
TRACE - deregistering handle with poller
DEBUG - dropping I/O source: 0
TRACE - signal: Closed
TRACE - event Readable Token(4194303)
TRACE - loop process - 1 events, 0.000s
TRACE - loop process - 0 events, 0.000s
TRACE - event Readable Token(4194303)
TRACE - loop process - 1 events, 0.000s
thread panicked while panicking. aborting.
error: process didn't exit successfully: `/home/otavio/src/updatehub/updatehub/target/debug/deps/updatehub-b92830f3badd740f 'tests::probe_'` (signal: 4, SIGILL: illegal instruction)
```

**Comment by lipanski** [comment]
@otavio could you share some code of how to reproduce this? (I mean I could infer it from your debug traces but maybe you have something simpler available)

**Comment by otavio** [comment]
You can use https://github.com/otavio/updatehub/; if you use it and upgrade to 0.15 it fails.

I tried to isolate it and could come with a test case but I am not sure it is the same error.

**Comment by arahalevich-move** [comment]
I get very similar error but using `Matcher::JsonString`. And it happens randomly. Say I have just one test with one mock with one request. I ran it 10 times and it fails 2 or 4 times out of 10.
It fails at `github.com-1ecc6299db9ec823/mockito-0.15.0/src/lib.rs:547`

**Comment by saks** [comment]
Minimal code to reproduce:
`Cargo.toml`:
```toml
[package]
name = "error_in_mockito"
version = "0.1.0"
authors = ["me <me@foo.com>"]
edition = "2018"

[dependencies]
reqwest = "0.9.9"
mockito = "0.15.0"
```

`src/lib.rs`:
```rust
pub fn get() {
    use reqwest::Client;
    let _ = Client::new()
        .post(&mockito::server_url())
        .body("{}")
        .send()
        .unwrap();
}

#[cfg(test)]
mod tests {
    use super::get;
    use mockito::{mock, Matcher};

    #[test]
    fn all() {
        let my_mock = mock("POST", "/")
            .with_status(200)
            .with_body("world")
            .match_body(Matcher::JsonString("{}".to_string()))
            .create();

        get();

        my_mock.assert();
    }
}
```

then run test multiple times:
`for i in {0..20}; do cargo t; done`
for me it fails several times

**Comment by lipanski** [comment]
@otavio Fix released as part of 0.15.1.
